compaction/time_window_compaction_strategy.cc

```diff
@@ -254,32 +269,53 @@ time_window_compaction_strategy::get_reshaping_job(std::vector<shared_sstable> i
         }
     }
 
-    auto is_disjoint = [&schema, mode, max_sstables] (const std::vector<shared_sstable>& ssts) {
+    auto is_disjoint = [&schema, mode, max_sstables = max_overlapping_sstables] (const std::vector<shared_sstable>& ssts) {
         size_t tolerance = (mode == reshape_mode::relaxed) ? max_sstables : 0;
         return sstable_set_overlapping_count(schema, ssts) <= tolerance;
     };
 
+    // Always respect off-strategy's space overhead goal.
+    // - With goal of 1.0 and disjoint input of N sstables, a single reshape will include all N.
+    // - With goal of 0.5 and disjoint input of N sstables, a single reshape will include N * 0.5.
+    // - With goal of 1.0 and overlapping input of N sstables, a single reshape will include min(N * 1.0, max_threshold)
+    // - With goal of 0.5 and overlapping input of N sstables, a single reshape will include min(N * 0.5, max_threshold)
+
+    auto get_max_sstables = [&] (const std::vector<shared_sstable>& ssts) {
```

Comment:
We can instead measure the amount of free space we have, and collect sstables that sum to not more than (free space - reserve) * 0.9. This way we avoid the extra configuration.

It's not so easy since free space is shared among shards. We can divide by smp::count, maybe that's sufficient.

With tablets, we don't have any problem, since even a 100% overhead results in 1% increase, assuming 100 tablets.

Reply:
We thought about that, but we decided for this simpler solution for backport. This series is already picking a default that cut overhead by half. But I can resurrect this idea if you prefer.

Reply:
I'd like to avoid adding yet another configuration (btw is it exposed in DESCRIBE SCHEMA?). Each configuration is one more thing that works or doesn't, and then gets tuned or doesn't.

Reply:
all compaction properties get exposed in describe schema, no recognition is needed for them. I can push for this more robust solution then. One of the reasons we didn't go for it is that it might theoretically take longer to implement and we weren't willing to spend too much time on this, because of what you said above, it won't be a problem with tablets. I share your sentiment with adding yet another tunable, adding another variable to our long option matrix. I don't think it's that much extra work, to be honest.